feat(artifacts): Enable GCE deploy stage to deploy an artifact

ezimanyi opened the pull request:
The deploy stage for the GCE provider now gives the option of deploying a specified upstream artifact rather than deploying the result of the most recent Bake or Find Artifact stage.

ezimanyi commented:
The changes here depend on the Orca changes in https://github.com/spinnaker/orca/pull/2049 to properly work; the change to Orca should be merged first.

sbwsg commented on app/scripts/modules/core/src/artifact/expectedArtifactSelector.component.ts:
```diff
@@ -23,21 +23,20 @@ class ExpectedArtifactSelectorComponent implements IComponentOptions {
             </div>
           </div>
-          <div class="form-group">
+          <div ng-if="ctrl.account-field" class="form-group">
```
should this be `accountField` to match the public prop name on line 10?

sbwsg commented on app/scripts/modules/google/src/help/gce.help.ts:
```diff
@@ -13,6 +13,16 @@ const helpContents: {[key: string]: string} = {
+  'gce.image.artifact': 'The artifact that is to be deployed to this cluster.  The artifact should represent a deployable image.',
+  'gce.image.source': `
+      <p>Where the image to deploy is read from.</p>
+            <p>
```
Indent looks wrong here

danielpeach commented on app/scripts/modules/core/src/artifact/imageSourceSelector.component.ts:
```diff
@@ -0,0 +1,52 @@
+/*
```
Please omit this. The build leaves comments in the generated app (or at least it did in the past), which we try to keep as small as possible.

danielpeach commented on app/scripts/modules/core/src/artifact/imageSourceSelector.component.ts:
```diff
@@ -0,0 +1,52 @@
+
+import { IComponentOptions, IController, module } from 'angular';
+
+class ImageSourceSelectorCtrl implements IController {
```
You can omit this controller entirely.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/stages/deploy/deployStage.js:
```diff
@@ -33,6 +33,7 @@ module.exports = angular.module('spinnaker.core.pipeline.stage.deployStage', [
             }
             return stage.clusters.every(cluster =>
               cloudProviderRegistryProvider.$get().getValue(cluster.provider, 'serverGroup.skipUpstreamStageCheck')
+              || cluster.imageSource === 'artifact'
```
This feels a little tacked on - might be better to include a flag in the the provider's config (like `skipUpstreamStageCheck`).

danielpeach commented on app/scripts/modules/core/src/serverGroup/configure/common/deployInitializer.component.ts:
```diff
@@ -77,13 +77,17 @@ export class DeployInitializerController implements IController {
     viewState.submitButtonLabel = 'Add';
     viewState.hideClusterNamePreview = baseCommand.viewState.hideClusterNamePreview || false;
     viewState.templatingEnabled = true;
+
```
Extraneous space.

ezimanyi commented:
@danielpeach : Addressed your comments; trivial changes in 'Next round of PR comments' and pulling out the validation into a function in 'Refactor deploy stage validation into a function'.  Let me know if you have any comments, particularly on the second one.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/stages/deploy/deployStage.js:
```diff
@@ -9,8 +10,9 @@ const angular = require('angular');
+  CLUSTER_SERVICE,
 ])
-  .config(function (pipelineConfigProvider, cloudProviderRegistryProvider) {
+  .config(function ($injector, pipelineConfigProvider, cloudProviderRegistryProvider, clusterServiceProvider) {
```
Are you using `$injector` for anything?

ezimanyi replied:
Nope, that's left over from when I was mis-understanding how to get access to the cluster service in the config function; I'll remove it.

danielpeach commented on app/scripts/modules/core/src/cluster/cluster.service.ts:
```diff
@@ -176,6 +176,10 @@ export class ClusterService {
   }
 
+  public isDeployingArtifact(cluster: ICluster): boolean {
+    return (cluster.imageSource === 'artifact');
```
The parens are valid JS here but not really idiomatic (you'll only see it if you're using React).

sbwsg commented on app/scripts/modules/kubernetes/src/v2/pipelines/stages/deployManifest/deployManifestConfig.html:
```diff
@@ -37,14 +37,16 @@ <h4>Manifest Configuration</h4>
-                              expected-artifacts="ctrl.expectedArtifacts"
-                              help-field-key="kubernetes.manifest.expectedArtifact">
-  </expected-artifact-selector>
+  <div class="container-fluid form-horizontal">
```
we'll end up with an empty `<div>` in the DOM if `ctrl.$scope.stage.source !== ctrl.artifactSource`.  Is this intentional or should the `ng-if` from `expected-artifact-selector` move up to this div?

ezimanyi replied:
Good point, I'll fix that

ezimanyi commented:
Screenshot of the new field in the "Basic Settings" section of "Configure Deployment Cluster"

![image](https://user-images.githubusercontent.com/1391982/37310758-54b6a02e-261b-11e8-9689-bc28b9daf681.png)
